### PR TITLE
Add proposal for simulation provider for testing

### PR DESCRIPTION
**Pull request by takoverflow**
**What this PR does / why we need it**:
This PR introduces a proposal to introduce a machine controller simulation provider to enhance developer experience by making manual testing easier and having IT run as part of MCM CI/CD pipelines without the need of real CSP infrastructure.

**Which issue(s) this PR fixes**:
Fixes part of #1120

**Special notes for your reviewer**:

**Release note**:

```other operator

```

### CODE REVIEW

File: docs/proposals/simulation-provider.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# Simulation Provider for Testing
```
Comment (ashwani2k) [inline docs/proposals/simulation-provider.md:1]: Won't Emulator be a better fit. As you emulate the cloud provider here, or may be call it virtual, I think simulation means you will still call the actual provider but you just simulate the tests locally. 
or may be this is a like a fake provider.

File: docs/proposals/simulation-provider.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  - controller_suite_test.go
+  - controller_test.go
+  - resource_tracker.go
+```
+
+#### High level Flow
+
+1. Initialise control plane i.e. start `kube-apiserver`, `etcd` as local processes.
+2. Deploy MCM CRDs.
+3. Start `machine-controller-manager` and `machine-controller-simulation-provider` processes.
```
Comment (ashwani2k) [inline docs/proposals/simulation-provider.md]: Even if we keep it as simulation we should still name it as `machine-controller-manager-provider-simulator|virtual|fake` to be inline with other providers.

File: docs/proposals/simulation-provider.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+4. Deploy user specified MCD and MCC (fallback to in-tree IT specific test MCD and MCC)
+5. Create dummy GNA secret (GNA secret, need to check if only passing a dummy environment variable for `GNA_SECRET_NAME` is sufficient or the secret itself needs to be created)
+6. Run IT. (Set some test related environment variables to true)
+
+## Target Scope
+
+Only offer functionality needed to run IT standalone (as part of CI/CD pipelines), with optional support for adding points of failures. The simulation config skaffold and supported modifications doesn't have to be exhaustive in the first iteration, however it should be extensible enough to allow someone to add their customizations to the simulation provider at their specified modification point to test edge cases (restricted still to modification related to driver interface calls).
```
Comment (ashwani2k) [inline docs/proposals/simulation-provider.md:147]: As currently we don't run IT standalone, it will be good to classify test added as part of this endeavor as mandatory vs non-mandatory for PR merge. Otherwise you might end up with IT test failing and never succeeding esp. if they validate scenarios which have time constraints.

**Review by ashwani2k — COMMENTED** [review]
Thanks for collating the thoughts here. Just some observations.

File: docs/proposals/simulation-provider.md (hunk 1 above)
Comment (elankath) [reply docs/proposals/simulation-provider.md:1]: @ashwani2k  It is exactly the other way around. 

> An emulator replicates both the hardware and software of a target device, acting as an exact clone (e.g., running an Android game on your PC). A simulator only mimics the software environment and its behavior, without replicating the underlying hardware (e.g., testing how an iOS app interface responds on your `Mac)

So, it is best to call it `simulated-provider`.

File: docs/proposals/simulation-provider.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+- Introduce modifications to the simulated machine's lifecycle in the form of delays, errors, etc.
+- Specify custom `MachineDeployment`s (MCD) and `MachineClass`es (MCC) for the end user to perform manual testing.
+
+## Proposal
+
+In the MCM repository tree, introduce a new package (tentatively named) `provider-simulation` which implements the `Driver` interface required by each supported cloudprovider. Refer to [machine_error_codes.md#machine-provider-interface](../development/machine_error_codes.md#machine-provider-interface) for more details.
+
+### Initialisation
+
+Cluster creation and lifecycle to be managed via `e2e-framework`, it handles the initialisation and creation of the minimal control plane needed for running the simulation provider. This fetches the specified `kube-apiserver` and `etcd` binaries (if not present) and runs them as local processes.
```
Comment (elankath) [inline docs/proposals/simulation-provider.md:18]: What is this `e2e-framework` ? Why can't we just use a `make target` for this ? `Ex: make test-integration-setup` ?

File: docs/proposals/simulation-provider.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+```json
+{
+  "preCreate": {
+    "minDelay": "10s",
+    "maxDelay": "30s",
+    "percentageOfMachines": "10%"
+  },
+  "preDelete": {
+    "rateLimitError": {
```
Comment (elankath) [inline docs/proposals/simulation-provider.md:59]: We need this configurable as a pattern for against MachineDeployments. ie we need a `machineDeploymentPattern` attribute somewhere in this json.

File: docs/proposals/simulation-provider.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  }
+}
+```
+
+### Integration with MCM
+
+The simulation provider would be added as a new package located at `pkg/simulator`.
+Proposed file structure:
+```
+pkg/simulator:
```
Comment (elankath) [inline docs/proposals/simulation-provider.md]: This should be `pkg/simulatedprovider`

File: docs/proposals/simulation-provider.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  - main.go
+  - Makefile
+  provider/
+  - driver.go
+  provider/simulation/
+  - simulation.go
+  - config.go
+  provider/simulation/injections/
+  - pre_create.go
```
Comment (elankath) [inline docs/proposals/simulation-provider.md]: I don't think files like `pre_create.go` are needed. What would even go inside this file ?  If someone wants to inject something outside of the MCM code base, then explicitly define an interface for this.

File: docs/proposals/simulation-provider.md (hunk 3 above)
Comment (takoverflow) [reply docs/proposals/simulation-provider.md:147]: Clarified with [381bde0](https://github.com/gardener/machine-controller-manager/pull/1121/commits/381bde0f2205e183b68610569a2e7e7c9b823fa9) PTAL

File: docs/proposals/simulation-provider.md (hunk 4 above)
Comment (takoverflow) [reply docs/proposals/simulation-provider.md:18]: As discussed offline, [kubernetes-sigs/e2e-framework](https://github.com/kubernetes-sigs/e2e-framework) is one of the recommended ways to create and consume environments for e2e k8s testing.

As part of the IT integration, its convenient helpers would be utilized for automated setup alongwith `make` targets for manual setup.

File: docs/proposals/simulation-provider.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+     ```
+   - Run `clustersim setup <demo-cluster> --mcc=./mcc.yaml --mcd=./mcd.yaml` with the fetched data; this would create a directory `./gen/<demo-cluster>/` containing the following files
+     ```
+     ./gen/<demo-cluster>
+     + start-config-mcm.json # Contains the launch flags for mcm
+     + start-config-mc.json # Contains the launch flags for mc-provider-simulation
+     + mcc.yaml
+     + mcd.yaml
+     ```
+   - Additionally the `setup` would also build the required binaries for `mcm` and `mc-provider-simulation` (in `gen/bin` directories, using their corresponding `make build` targets) which can later be used when running the testing environment.
```
Comment (thiyyakat) [inline docs/proposals/simulation-provider.md:126]:
```suggestion
   - Additionally the `setup` would also build the required binaries for `mcm` and `mcm-provider-simulator` (in `gen/bin` directories, using their corresponding `make build` targets) which can later be used when running the testing environment.
```
Is this what you meant to write?

**Review by thiyyakat — COMMENTED** [review]
Thanks for creating the proposal! Just one nit.
